### PR TITLE
[SEDONA-625] Add ST_GeneratePoints

### PR DESCRIPTION
**Pull request by furqaankhan**
## Did you read the Contributor Guide?

- Yes, I have read the [Contributor Rules](https://sedona.apache.org/latest-snapshot/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/latest-snapshot/community/develop/)

## Is this PR related to a JIRA ticket?

- Yes, the URL of the associated JIRA ticket is https://issues.apache.org/jira/browse/SEDONA-XXX. The PR name follows the format `[SEDONA-XXX] my subject`.

## What changes were proposed in this PR?

- Add ST_GeneratePoints

## How was this patch tested?

- Add new unit tests

## Did this PR include necessary documentation updates?

- Yes, I have updated the documentation.

### CODE REVIEW

**Review by jiayuasu — APPROVED** [review]

**Comment by jiayuasu** [comment]
@Kontinuation The result of this function is non-deterministic. Will this create a problem to the catalyst optimizer?

**Comment by Kontinuation** [comment]
> @Kontinuation The result of this function is non-deterministic. Will this create a problem to the catalyst optimizer?

The correct way to implement non-deterministic functions is to extend the `Nondeterministic` trait and implement the `initializeInternal` method to calculate the seed for each partition. The function evaluates using the initialized seed, so the results are identical when partitions are recomputed.

Unfortunately, `RandomPointsBuilder` does not allow constructing the builder with a given seed. We have to inherit this class and override the `createRandomCoord` function to generate coordinates using the specified seed.

**Comment by jiayuasu** [comment]
@furqaankhan can you solve the issue using Kristin's method?

**Comment by furqaankhan** [comment]
Yes, I am working on it.

**Comment by jiayuasu** [comment]
This PR should also add SRID to the generated points following the style of this PR: https://github.com/apache/sedona/pull/1521

File: spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala [hunk 1]
```diff
@@ -1437,6 +1439,60 @@ case class ST_ForceRHR(inputExpressions: Seq[Expression])
...
+
+  override protected def evalInternal(input: InternalRow): Any = {
+    val geom = children.head.toGeometry(input)
+    val numPoints = children(1).eval(input).asInstanceOf[Int]
+    if (nArgs == 3) {
+      val seed = children(2).eval(input).asInstanceOf[Int]
+      return GeometrySerializer.serialize(Functions.generatePoints(geom, numPoints, seed))
+    }
+    GeometrySerializer.serialize(Functions.generatePoints(geom, numPoints))
+  }
```
Comment (Kontinuation) [inline spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala:1472]: We should construct an instance of `RandomPointsBuilderSeed` in the `initializeInternal` function, and call    `setExtent`, `setNumPoints`, `getGeometry` methods using that instance in `evalInternal`.

File: spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala [hunk 2]
```diff
@@ -1437,6 +1439,60 @@ case class ST_ForceRHR(inputExpressions: Seq[Expression])
...
 }
 
+case class ST_GeneratePoints(inputExpressions: Seq[Expression], randomSeed: Option[Long] = None)
+    extends Expression
+    with Nondeterministic
+    with ExpectsInputTypes
+    with CodegenFallback
+    with ExpressionWithRandomSeed {
+
+  def this(inputExpressions: Seq[Expression]) = this(inputExpressions, Some(0L))
```
Comment (Kontinuation) [inline spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala:1449]: `ExpressionWithRandomSeed` has binary incompatible changes since Spark 3.0, so we cannot use it to write binary compatible code for Spark 3.0 ~ 3.5. Maintaining compatibility with old Spark versions (<= 3.2) is really a burden for us.

We can initialize `ST_GeneratePoints` in the old Spark 2.3 way:

```
case class ST_GeneratePoints(inputExpressions: Seq[Expression], randomSeed: Long) extends RDG {

  def this(inputExpressions: Seq[Expression]) = this(inputExpressions, Utils.random.nextLong())
```

The consequence is that the random seed will be fixed for multiple minibatches when running streaming jobs. However, that's the best we can do to maintain compatibility with various Spark versions.

Comment (furqaankhan) [reply spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala]: Using RDG doesn't work, as extends `UnaryExpression`.

**Comment by furqaankhan** [comment]
> This PR should also add SRID to the generated points following the style of this PR: #1521

I am using the factory of the input geometry. that will preserve the SRID. Please let me know if you think otherwise.

https://github.com/furqaankhan/sedona/blob/919136e2d67e730bd864f87c466a7f8be3b84407/common/src/main/java/org/apache/sedona/common/Functions.java#L1808

File: spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala (hunk 2 above)
Comment (Kontinuation) [reply spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala]: Ah, I made a mistake when writing the example code. It should be

```
case class ST_GeneratePoints(inputExpressions: Seq[Expression], randomSeed: Long) extends Expression
    with Nondeterministic
    with ExpectsInputTypes
    with CodegenFallback  {

  def this(inputExpressions: Seq[Expression]) = this(inputExpressions, Utils.random.nextLong())
}
```

Comment (furqaankhan) [reply spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala]: I was going to push this changes by removing the `ExpressionWithRandomSeed` trait and it's overridden members, to test it.

**Comment by jiayuasu** [comment]
@furqaankhan Can you create a PR on snowflake-tester?

**Comment by furqaankhan** [comment]
I did create one, I closed it as it had passed. I  reopened it as I added a new function signature. [The PR.](https://github.com/wherobots/sedona-snowflake-tester/pull/27)

File: common/src/main/java/org/apache/sedona/common/Functions.java [hunk 3]
```diff
@@ -1812,6 +1812,24 @@ private static Geometry[] convertGeometryToArray(Geometry geom) {
...
+    pointsBuilder.setExtent(geom);
+    pointsBuilder.setNumPoints(numPoints);
+    return pointsBuilder.getGeometry();
+  }
+
+  public static Geometry generatePoints(Geometry geom, int numPoints) {
+    return generatePoints(geom, numPoints, 0);
+  }
+
+  public static Geometry generatePoints(Geometry geom, int numPoints, Random random) {
```
Comment (jiayuasu) [inline common/src/main/java/org/apache/sedona/common/Functions.java:1826]: Why is there another overloaded function that takes `Random` as input?

File: spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala [hunk 4]
```diff
@@ -1435,6 +1436,58 @@ case class ST_ForceRHR(inputExpressions: Seq[Expression])
...
+
+  override protected def evalInternal(input: InternalRow): Any = {
+    val geom = children.head.toGeometry(input)
+    val numPoints = children(1).eval(input).asInstanceOf[Int]
+    val generatedPoints = if (nArgs == 3) {
+      val seed = children(2).eval(input).asInstanceOf[Int]
+      if (seed > 0) {
+        Functions.generatePoints(geom, numPoints, seed)
+      } else {
+        Functions.generatePoints(geom, numPoints, random)
```
Comment (jiayuasu) [inline spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala:1462]: Why is this needed? This if condition seems to be repeating the same condition in `Functions`

**Review by jiayuasu — CHANGES_REQUESTED** [review]

File: common/src/main/java/org/apache/sedona/common/Functions.java (hunk 3 above)
Comment (Kontinuation) [reply common/src/main/java/org/apache/sedona/common/Functions.java:1826]: This is for using the `random` member of `ST_GeneratePoints` to generate consistent random streams

File: spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala (hunk 4 above)
Comment (Kontinuation) [reply spark/common/src/main/scala/org/apache/spark/sql/sedona_sql/expressions/Functions.scala:1462]: This is for generating consistent random streams even if seed is 0.

**Review by jiayuasu — APPROVED** [review]
